### ISSUE TITLE
do syntax for array constructors

### ISSUE THREAD

**Issue by ExpandingMan** [JuliaLang/julia#32435]
I thought me or someone else had already suggested this, but I can't find the issue.

A very common pattern is to create an undefined array and iterate over it with a `for` loop to fill it
```julia
v = Array{Float64}(undef, (4,4,4))   
for idx ∈ CartesianIndices(v)
    # do stuff
    v[idx] = something(idx)
end
```

I think it would be really nice if there were `do` block syntax for this such as
```julia
v = Array{Float64}(undef, (4,4,4)) do idx
    # do stuff 
    something(idx)
end
```

This should only require adding some methods to `Array`.  One could add additional methods to allow the user to specify whether he wants `CartesianIndices` or `LinearIndices`.

Any interest in this?

**Comment by yuyichao** [comment]
Just use a comprehension?

**Comment by ExpandingMan** [comment]
This would be useful in cases where your code is big enough to make a comprehension awkward.

**Comment by yuyichao** [comment]
Well, so you just want an alternative syntax to comprehension? If you think the comprehension syntax is awkward sometimes, then I think it would also be awkward for the cases not covered by your function. In that case, the alternative syntax had better have the same function as the original one. However, what you currently have is basically a strict subset of what comprehension could do (and can easily be defined as such `(::Type{Array})(f, idxs) = [f(idx) for idx in idxs]; (::Type{Array{T}})(f, idxs) where T = T[f(idx) for idx in idxs]`).

Although I'm not entirely sure if I agree with that the comprehension syntax is awkward for long body, we already have a generic solution for this that works not only for comprehension but for any code blocks. You can just define a function explicitly,

```julia
f = function(idx)
    # Do whatever you want.
end
ary = [f(idx) for idx in idxs]
```

**Comment by ExpandingMan** [comment]
Yes, I agree that this would be equivalent to comprehension defined over a function.  That this could be easily defined they way you suggested is sort of my point: it would be a very simple addition that would make it just a little more convenient to perform an extremely common pattern, and it does not require any actual new syntax.  In my view this is a bit like 1-line function definitions: yes, you could easily define it in a block, but it's extremely common to want to define functions in one line, so why not allow that syntax.

I think my suggestion would be silly were this not an extremely commonly used pattern, it's the ubiquitousness of the pattern that makes me think it might be worth it.

**Comment by dkarrasch** [comment]
In such cases, I'm using
```julia
v = map(idxs) do idx
    # do stuff 
    return something(idx)
end
```
which has the `do`-block style, doesn't force you to stick to the actual indices of the resulting `Array`, and also avoids the confusing `undef` in the OP.

**Comment by ExpandingMan** [comment]
That's a good point, I feel stupid for not having thought of using `map` that way.  However, I personally probably still would not use that most of the time, because it doesn't allow you to explicitly declare the element type in the array (though you could put a type assertion or `convert` on the return values).
